### ISSUE TITLE
compiler raising Invalid_argument("output_value: functional value")

### ISSUE THREAD

**Issue by craff** [ocaml/ocaml#11387]
Hello,

When using dune (I was not able to reproduce with simple compilation command), ocamlc.opt raises the mentioned exception.

I took the time to produce a minimal example (13 lines) triggering the bug. It is related to module type constraints.

I tested with 4.14.0 and 4.14.0+flambda

to reproduce, just untar the archive and execute dune build inside 

dune build --verbose produces:

$ (cd _build/default && /home/raffalli/.opam/4.14.0/bin/ocamlc.opt -w +a-70-42-44-45-4 -warn-error +a -g -bin-annot -I lib/.T2.objs/byte -no-alias-deps -opaque -open T2__ -o lib/.T2.objs/byte/t2.cmo -c -impl lib/T2.ml)
Fatal error: exception Invalid_argument("output_value: functional value")

[Bug.tar.gz](https://github.com/ocaml/ocaml/files/9033361/Bug.tar.gz)

**Comment by kit-ty-kate** [comment]
Using your example I was able to find a smaller reproduction case with the current `5.0` branch (https://github.com/ocaml/ocaml/commit/02bdcf2d8e056da53cac9d0fc26cf0eb9c7536c9):

`test.ml`:
```
module T = struct
  module type T = sig module type M end
end

open T

module A = struct
  module B(T:T) : T with module type M = T.M = struct
    include T
  end
end
```
```
$ ocamlc -w +45 -bin-annot test.ml
Fatal error: exception Invalid_argument("output_value: functional value")
```
Note that without both `-w +45` and `-bin-annot` the bug does not appear.

**Comment by craff** [comment]
I did not try to play with -w when trying to reproduce without dune! strange that a -w flag can produce such a bug.

**Comment by Octachron** [comment]
The issue is due to the environment in the helper node `Twith_modtype` which is not cleaned of functions. This is why it requires both a warning and a `with` module type constraint. Fix incoming.

**Comment by xavierleroy** [comment]
Looks like this was fixed in the trunk, 5.0 and  4.14 branches.
